### ISSUE TITLE
[Epic] Implement support for `StringView` in DataFusion

### ISSUE THREAD

**Issue by alamb** [apache/datafusion#10918]
### Is your feature request related to a problem or challenge?

`StringView` / `BinaryView` were added to the Arrow format that make it more suitable for certain types of operations on strings.  Specifically when filtering with string data, creating the output `StringArray` requires copying the strings to a new, packed binary buffer. 

[`GenericByteViewArrary`](https://docs.rs/arrow/latest/arrow/array/struct.GenericByteViewArray.html) was designed to solve this limitation and the arrow-rs implementation, tracked by https://github.com/apache/arrow-rs/issues/5374, is now complete enough to start adding support into DataFusion

I think we can improve performance in certain cases by using StringView (this is also described in more details in [the Pola.rs blog post](https://pola.rs/posts/polars-string-type/))
1. Reading strings / binary from Parquet files as `StringViewArray`/`BinaryViewArray`  rather than `StringArray` / `BinaryArray` saves a copy (and @ariesdevil  is quite close to having it integrated into the parquet reader https://github.com/apache/arrow-rs/issues/5530) 
2. Evaluating predicates on string expressions (for example `substr(url, 4) = 'http'`) as the intermediate result of `substr` can be called without copying string values

### Describe the solution you'd like
I would like to support StringView / BinaryView support in DataFusion.

While my primary usecase is for reading data from parquet, I think teaching DataFusion to use `StringView` (at least as intermediate values when evaluating expressions may help significantly

### Development branch: `string-view`

Since this feature requires  upstream arrow-rs support https://github.com/apache/arrow-rs/issues/5374 that is not yet released we plan to do development on a `string-view` feature branch:

https://github.com/apache/datafusion/tree/string-view

### Task List
Here are some high level tasks (I can help flesh these out for anyone who is interested in helping)
- [x] https://github.com/apache/datafusion/issues/10920
- [x] https://github.com/apache/datafusion/issues/10919
- [x] https://github.com/apache/datafusion/issues/10998
- [x] https://github.com/apache/datafusion/issues/11024
- [x] #10996
- [x] https://github.com/apache/datafusion/issues/10921
- [x] https://github.com/apache/datafusion/issues/10961
- [x] https://github.com/apache/datafusion/issues/11024
- [x]  https://github.com/apache/datafusion/issues/11025
- [x] #11023
- [x] https://github.com/apache/arrow-rs/pull/5936
- [x] Add generic coercion rule from `Utf8View` -> `Utf8` / `BinaryView` -> `Binary` for compatibility
- [ ] File tickets for adding native Utf8View / BinaryView support to other functions which would benefit
- [ ] Establish pattern for implementing optimized string functions for StringViewArray
- [ ] https://github.com/apache/datafusion/issues/11682
- [x] https://github.com/apache/datafusion/issues/11677

### Additional context

Polars implemented it recently in rust so that can serve as a motivation
Blog Post https://pola.rs/posts/polars-string-type/
https://twitter.com/RitchieVink/status/1749466861069115790

Facebook/Velox's take: https://engineering.fb.com/2024/02/20/developer-tools/velox-apache-arrow-15-composable-data-management/

Related PRs:
https://github.com/pola-rs/polars/pull/13748
https://github.com/pola-rs/polars/pull/13839
https://github.com/pola-rs/polars/pull/13489

**Comment by alamb** [comment]
I think we should aim for a first "milestone" of showing improvements for some clickbench queries

**Comment by jayzhan211** [comment]
Will we completely change StringArray to StringViewArray in Datafusion?
While I try to utilize StringViewArray in #10976 , I found there is schema mismatched issue UTF8 vs UTF8View. To avoid converting StringViewArray to StringArray, we might need to change the schema to string view overall from logical plan to physical plan. If we need to keep both String and StringView, then we need to think about how to deal with the conversion between these two types.

**Comment by alamb** [comment]
> Will we completely change StringArray to StringViewArray in Datafusion?

I think since they are two separate types in Arrow we couldn't fully switch to StringView the way polars could as it controls the whole stack. Users could still feed DataFusion StringViewArray from custom TableProviders and would expect StringView at the output. 

However what I think we *could* do is internally to DataFusion (e.g. within the plan, before the final output) is use StringView in the batches that flow through intermediate nodes in the plan.

>  I found there is schema mismatched issue UTF8 vs UTF8View. To avoid converting StringViewArray to StringArray, we might need to change the schema to UTF8View 

Indeed, As you point out, I don't think we can transparently switch to using StringView -- instead we would have to start encoding information in the plans about the new types.

I wonder if we could have a new logical optimzier pass that tried to annotate operations that support it to use StringView in their schema rather than String. Then the ExecutionPlans would know if they were supposed to generate StringView as output or the more traditional StringArray 🤔

**Comment by alamb** [comment]
I think @XiangpengHao  is looking into another place to use StringView which is https://github.com/apache/datafusion/issues/10921 -- where we have a similar idea to use StringView in some sub portion of the plan. Here is more info about the optimizer pass idea: https://github.com/apache/datafusion/issues/10921#issuecomment-2178364966

**Comment by alamb** [comment]
I think this project is going pretty well :bowtie: 

We are at the point of starting to implement some basic functions using StringView.

- [ ] https://github.com/apache/datafusion/issues/11024
- [ ] https://github.com/apache/datafusion/issues/11025

**Comment by alamb** [comment]
Now that we have upgraded to arrow 52.1.0, I think we could merge the `string-view` branch to main. I'll try and make a PR if no one beats me to it

**Comment by alamb** [comment]
An update here on the plan from @XiangpengHao:
* He has local changes (that also require some additional features that will be released in arrow `52.2.0`) that show significant performance improvements for TPCH and ClickBench queries
* The hope is that these proposals are all up for review by the end of this week 
* So by the time arrow `52.2.0` is released (early August 2024) we'll be able to add an option that makes DataFusion use StringView when reading from Parquet / filtering
* We also plan to write a blog post about this work / adventure

I for one am very excited

**Comment by alamb** [comment]
Update: we merged the basic support for StringView in  https://github.com/apache/datafusion/pull/11402

I have created a branch to collect any changes that rely on the pre-release parquet/arrow `52.2.0` version here: https://github.com/apache/datafusion/tree/string-view2

**Comment by alamb** [comment]
Update here is that we hope to have enough code merged / implemented to report benchmark results by the end of the week. Once arrow-rs 52.2.0 is released https://github.com/apache/arrow-rs/issues/5998 then we'll merge the contents of https://github.com/apache/datafusion/tree/string-view2 to main

Once we do that we'll evaluate what, if anything, is left before we can close this ticket

**Comment by alamb** [comment]
Update here is that we have all the basic support in now but it is behind a feature flag. Let's declare this project a success for now and track the remaining work to enable this feature by default in a separate issue: https://github.com/apache/datafusion/issues/11752

Thanks to everyone who has helped so far. We are close ™️

**Comment by alamb** [comment]
In case anyone wants an overview of adding StringView to DataFusion, here is a presentation and slides from @XiangpengHao 
Video: https://www.youtube.com/watch?v=RVLshX6fbds
Slides: https://drive.google.com/file/d/1Qqd8V6cfS9rSQ_-JrinasQJwI79qlUEV/view?usp=drive_link
